docs(Contributing.md): Add Contributing.md

M-0-N-K opened the pull request:
docs ( Contributing.md): added Contributing.md

M-0-N-K committed:
Create Contribution.md

chirgjn commented:
Let's use this convention instead:
https://www.conventionalcommits.org/en/v1.0.0/

Can you mention the steps to use the following CLI, so that commit messages are as per convention.
http://commitizen.github.io/cz-cli/

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+    3. Shared types should be defined in `types.ts`.
+    4. Within a file, type definitions should come first.
+
+  ### `null` and `undefined`
+    1. Use `undefined`. Do not use null.
```
What's the idea behind this?

M-0-N-K replied:
Sometimes there are functions that accept parameters as null or object and although we could send a null in some variables where we control the flow, sometimes other code might send undefined or simply not send anything thus making the parameters undefined. This makes us do a check for both, param==null || param == undefined . We can avoid the extra check for null.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+The guidelines for contribution to babitz backend project are listed below. Please go through it before any contribution.
+
+## 1. General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority. 
```
Please link Loopback 4's guidelines.

M-0-N-K replied:
done.

chirgjn replied:
I can't find the link even in the latest commit.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+## 1. General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority. 
+
+  ### Names
```
Can you link to the respective sections of the TypeScript contribution guidelines.

M-0-N-K replied:
done.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+    1. Do not use `for..in` statements; instead, use `ts.forEach`, `ts.forEachKey` and `ts.forEachValue`. Be aware of their slightly different semantics.
+    2. Try to use `ts.forEach`, `ts.map`, and `ts.filter` instead of loops when it is not strongly inconvenient.
+
+  ### Style
+    1. Use arrow functions over anonymous function expressions.
+    2. Only surround arrow function parameters when necessary. <br />For example, `(x) => x + x` is wrong but the following are correct:
+        - `x => x + x`
+        - `(x,y) => x + y`
+        - `<T>(x: T, y: T) => x === y`
+    3. Always surround loop and conditional bodies with curly braces. Statements on the same line are allowed to omit braces.
+    4. Open curly braces always go on the same line as whatever necessitates them.
+    5. Parenthesized constructs should have no surrounding whitespace. <br />A single space follows commas, colons, and semicolons in those constructs. For example:
+        - `for (var i = 0, n = str.length; i < 10; i++) { }`
+        - `if (x < 10) { }`
+        - `function f(x: number, y: string): void { }`
+    6. Use a single declaration per variable statement <br />(i.e. use `var x = 1; var y = 2;` over `var x = 1, y = 2;`).
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
```
We are using `prettier` to take enforce style. Let's link the instruction of running the same here.

M-0-N-K replied:
done.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
+
+## Loopback4 Guidelines
```
```suggestion
## File naming convention
```

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+
+## How to Raise Pull Requests 
+- Avoid making large pull requests so that code review gets easier.
+- Use either one feature/fix or one commit per pull request only.
```
Please describe trunk based development.
Please add branch name examples.

M-0-N-K replied:
done.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+
+- Write tests in the __test__ folder.
+- Make sure you cover at least 50% of the test-case scenario.
+- You can generate "test coverage" using the `npm run test:coverage` command.
```
Add references for how to write tests and what tools are used for testing.

chirgjn commented on Contribution.md:
```diff
@@ -0,0 +1,153 @@
+- We clearly define our api spec before coding the models and endpoints for them. Please have a look at how to [import api spec in loopback 4.](https://loopback.io/doc/en/lb4/OpenAPI-generator.html)
+
+
+## Directory structure
```
Add relevant reference: https://loopback.io/doc/en/lb4/Loopback-application-layout.html

chirgjn replied:
This is still pending

chirgjn commented:
https://github.com/github/docs/blob/main/CONTRIBUTING.md

M-0-N-K committed:
docs(contributing.md): change commit guidelines. Use commitzen's conventional changelog adapter

chirgjn commented:
@M-0-N-K a lot of the comments are still un-resolved

M-0-N-K committed:
docs (Contributing) : update Contributing.md 


- Add link to typescript's guideline sections.
- Add instructions for running prettier.
- Add guidelines on how to write test-cases.

chirgjn commented on Contributing.md:
```diff
@@ -139,15 +140,17 @@ Commit messages should adhere to the following guidelines:
-## How to Raise Pull Requests 
-- Avoid making large pull requests so that code review gets easier.
+## How to Raise Pull Requests
+- We use trunk based developement for our developement. Please use the same.
```
Please add a link for "trunk based development"

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,156 @@
+
+Reference: [how to write good commits](https://www.conventionalcommits.org/en/v1.0.0/)
+Please use [commitizen](https://www.npmjs.com/package/commitizen) to format your commit messages.
+steps to use commitizen:
+- Install commitizen globally with, "npm install commitizen -g"
+- Setup commitizen's conventional-changelog adapter in your local setup with "commitizen init cz-conventional-changelog --yarn --dev --exact"
```
We have already linked to the CLI documentation, hence this is redundant.
```suggestion
```

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,156 @@
+- Instructions on [how to use prettier cli.](https://prettier.io/docs/en/cli.html)
+
+
+## [Testing](https://loopback.io/doc/en/lb4/Testing-your-application.html)
```
Add a point linking to Loopback 4 testing best practices

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,156 @@
+    1. Do not use `for..in` statements; instead, use `ts.forEach`, `ts.forEachKey` and `ts.forEachValue`. Be aware of their slightly different semantics.
+    2. Try to use `ts.forEach`, `ts.map`, and `ts.filter` instead of loops when it is not strongly inconvenient.
+
+  ### [Style](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#style)
+    1. Use arrow functions over anonymous function expressions.
+    2. Only surround arrow function parameters when necessary. <br />For example, `(x) => x + x` is wrong but the following are correct:
+        - `x => x + x`
+        - `(x,y) => x + y`
+        - `<T>(x: T, y: T) => x === y`
+    3. Always surround loop and conditional bodies with curly braces. Statements on the same line are allowed to omit braces.
+    4. Open curly braces always go on the same line as whatever necessitates them.
+    5. Parenthesized constructs should have no surrounding whitespace. <br />A single space follows commas, colons, and semicolons in those constructs. For example:
+        - `for (var i = 0, n = str.length; i < 10; i++) { }`
+        - `if (x < 10) { }`
+        - `function f(x: number, y: string): void { }`
+    6. Use a single declaration per variable statement <br />(i.e. use `var x = 1; var y = 2;` over `var x = 1, y = 2;`).
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
```
We don't need these guidelines as we are using `prettier` for styling.
```suggestion
```

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,156 @@
+
+## Linting and Formatting
+ Apart from the typescript [Style](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#style) we use prettier and ESLint.
+- Use ESLint to statically analyse the source code and detect common problems. Config used from loopback/eslint-config.
+- Prettier to keep our code always formatted the same way. Config stored in .prettierrc
+- Both of these can be done with `  npm run lint:fix  `
+- Instructions on [how to use prettier cli.](https://prettier.io/docs/en/cli.html)
```
Both prettier and ESLint will be installed as part of the project, let's only link to their home page and document the various "check" and "fix' commands here.

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,156 @@
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
+
+## [Loopback4 Guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines)
```
The link here is for TS guidelines

chirgjn replied:
Let's rename this to "File naming convention"

M-0-N-K committed:
Update Contributing.md

Co-authored-by: Chirag Jain <chirgjn@users.noreply.github.com>

M-0-N-K committed:
Update Contributing.md

Co-authored-by: Chirag Jain <chirgjn@users.noreply.github.com>

M-0-N-K committed:
Docs (Contributing.md) Update Contributing.md according to feedback.

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,135 @@
+## File naming convention
+- File Naming Convention taken from [angular's file naming convention](https://angular.io/guide/styleguide#separate-file-names-with-dots-and-dashes). Loopback internally uses this so this decision is taken to have consistency in file names. This states we use a file name to also represent what it does or is.
+    - an example is "user.model.ts" here we can see the file is a "model" and the name of the model is "user".
+- We clearly define our api spec before coding the models and endpoints for them. Please have a look at how to [import api spec in loopback 4.](https://loopback.io/doc/en/lb4/OpenAPI-generator.html)
```
Add a separate heading for API spec

chirgjn commented on Contributing.md:
```diff
@@ -0,0 +1,135 @@
+The guidelines for contribution to babitz backend project are listed below. Please go through it before any contribution.
+
+## General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority.
```
Please link to Loopback 4 guidelines

M-0-N-K replied:
done.

chirgjn commented:
@M-0-N-K *ping*
